python/opae.admin/opae/admin/tools/rsu.py

```diff
@@ -47,48 +47,115 @@
-           FPGA image will be reconfigured from the user bank.
-
-     %(prog)s nextboot 25:00.0 --fpga=1
-     This sets the FPGA image to load on the next boot of the device with
-     PCIe address 25:00.0 to the FPGA User1 image.
+     This will trigger a factory boot of the BMC image for the device
+     with PCIe address 25:00.0.
+     NOTE: The BMC image will be reconfigured from factory bank and the
+           FPGA image will be reconfigured using the default setting.
+
+     %(prog)s fpgadefault 25:00.0 --page=factory --fallback=user1,user2
+     This sets the default FPGA image of the device with PCIe address
+     25:00.0 to the FPGA Factory image, with a fallback to FPGA user1
+     then FPGA user2.
 '''
 
 
-def bmc_available_image(args):
-    return 'bmc_' + args.page
+def fpga_defaults_valid(pci_id, value):
+    sequences = { (0x8086, 0xaf00): [ 'fpga_user1',
+                                      'fpga_user2',
+                                      'fpga_factory',
+                                      'fpga_factory fpga_user1',
+                                      'fpga_factory fpga_user2',
+                                      'fpga_factory fpga_user1 fpga_user2',
+                                      'fpga_factory fpga_user2 fpga_user1'
+                                    ],
+                  (0x8086, 0xaf01): [ 'fpga_user1',
+                                      'fpga_user2',
+                                      'fpga_factory',
+                                      'fpga_factory fpga_user1',
+                                      'fpga_factory fpga_user2',
+                                      'fpga_factory fpga_user1 fpga_user2',
+                                      'fpga_factory fpga_user2 fpga_user1'
+                                    ]
+                }
+    return value in sequences[pci_id]
 
 
-def retimer_available_image(args):
-    return 'retimer_fw'
+def set_fpga_default(device, args):
+    security = device.security
+    if not security:
+        logging.error('failed to find secure '
+                      'attributes for {}'.format(device.pci_node.pci_address))
+        raise IOError
 
+    power_on_image = security.find_one('power_on_image')
 
-def fpga_available_image(args):
-    images = {
-        '1': 'user1',
-        '2': 'user2',
-        'factory': 'factory'
-    }
-    return 'fpga_' + images[args.page]
+    if not args.page and not args.fallback:
+        # Print the power_on_image value.
+        logging.info('fallback sequence: {}'.format(power_on_image.value))
+        return
+    elif not args.page and args.fallback:
+        logging.error('--fallback must be accompanied by a --page selection.')
+        raise IOError
+
+    values = ['fpga_' + args.page]
+
+    if args.fallback:
+        fb = args.fallback
+        fb = fb.replace(',', ' ')  # allow a comma-separated list
+        while fb.find('  ') >= 0:
+            fb = fb.replace('  ', ' ') # compress spaces
```

Comment:
Can this be done with re.sub?
Something like:
```Python
fb = re.sub('\s+',' ', fb)
```

Reply:
yep